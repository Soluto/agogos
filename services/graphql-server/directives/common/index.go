package common

import "agogos/directives/middlewares"

// MiddlewareDefinitions common
var MiddlewareDefinitions = getMiddlewareDefinitions()

func getMiddlewareDefinitions() map[string]middlewares.MiddlewareDefinition {
	directives := make(map[string]middlewares.MiddlewareDefinition)
	directives["log"] = log
	directives["overrideContext"] = overrideContext
	directives["stub"] = stub
	directives["http"] = httpMiddleware
	directives["select"] = selectMiddleware
	directives["alias"] = aliasMiddleware
	return directives
}
