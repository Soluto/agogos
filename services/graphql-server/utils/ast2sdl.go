package utils

import (
	"fmt"
	"reflect"
	"strings"

	"github.com/graphql-go/graphql"
	"github.com/graphql-go/graphql/language/ast"
	"github.com/sirupsen/logrus"
)

// GqlRequestConfig - type to store all data for gql request
type GqlRequestConfig struct {
	Query          string
	VariableValues map[string]interface{}
}

type sdlQuery struct {
	builder        *strings.Builder
	fragmentsToAdd []string
	addedFragments map[string]bool
	variables      map[string]string
}

// ResolveParamsToSDLRequest - generates SDL query and variables from graphql.ResolveParams object
func ResolveParamsToSDLRequest(queryName string, rp graphql.ResolveParams, args string) GqlRequestConfig {
	query := sdlQuery{
		builder:        &strings.Builder{},
		addedFragments: make(map[string]bool),
		variables:      make(map[string]string),
	}

	query.builder.WriteString("{\n")
	query.builder.WriteString(queryName)
	query.builder.WriteString(" ")

	buildCustomArgumentsClause(query.builder, rp, args)

	for _, field := range rp.Info.FieldASTs {
		resolveParamsToQueryInner(&query, field, rp)
	}
	query.builder.WriteString("}\n")

	buildFragmentsClause(&query, rp)

	variablesClause := buildVariableClause(&query)

	return GqlRequestConfig{
		Query:          fmt.Sprintf("query%s%s", variablesClause, query.builder.String()),
		VariableValues: getVariableValues(query.variables, rp.Info.VariableValues),
	}

}

func resolveParamsToQueryInner(query *sdlQuery, definition ast.Selection, rp graphql.ResolveParams) {
	selectionSet := definition.GetSelectionSet()

	if selectionSet == nil || len(selectionSet.Selections) == 0 {
		query.builder.WriteString("\n")
		return
	}

	if field, ok := definition.(*ast.Field); ok {
		buildArgumentsClause(query, field, rp)
	}

	query.builder.WriteString("{\n")
	for _, selection := range selectionSet.Selections {
		// TODO: filter out fields with own resolvers

		switch selection.(type) {
		case *ast.Field:
			subField := selection.(*ast.Field)
			if subField.Alias != nil {
				query.builder.WriteString(subField.Alias.Value)
				query.builder.WriteString(": ")
			}

			query.builder.WriteString(subField.Name.Value)
			resolveParamsToQueryInner(query, subField, rp)

		case *ast.FragmentSpread:
			fragmentSpread := selection.(*ast.FragmentSpread)
			query.builder.WriteString(fmt.Sprintf("...%s\n", fragmentSpread.Name.Value))
			if addedToQuery, ok := query.addedFragments[fragmentSpread.Name.Value]; !ok || !addedToQuery {
				query.fragmentsToAdd = append(query.fragmentsToAdd, fragmentSpread.Name.Value)
			}
		default:
			logrus.WithField("selectionType", reflect.TypeOf(selection)).Panic("Unknown selection type")
		}
	}
	query.builder.WriteString("}\n")
}

func buildArgumentsClause(query *sdlQuery, field *ast.Field, rp graphql.ResolveParams) {
	if field.Arguments != nil && len(field.Arguments) > 0 {
		query.builder.WriteString("(")

		for _, arg := range field.Arguments {
			var argumentValue string
			switch arg.Value.GetKind() {
			case "StringValue":
				argumentValue = fmt.Sprintf("\"%s\"", arg.Value.GetValue())
			case "Variable":
				varName := arg.Name.Value
				if _, ok := query.variables[varName]; !ok {
					query.variables[varName] = getVariableType(varName, rp)
				}
				argumentValue = fmt.Sprintf("$%v", arg.Name.Value)
			default:
				argumentValue = fmt.Sprintf("%v", arg.Value.GetValue())
			}
			query.builder.WriteString(fmt.Sprintf("%s:%s", arg.Name.Value, argumentValue))
		}

		query.builder.WriteString(")")
	}
}

func buildCustomArgumentsClause(builder *strings.Builder, rp graphql.ResolveParams, args string) {
	if args == "" {
		return
	}

	builder.WriteString("(")
	values := ReplaceWithParameters(rp, args)
	builder.WriteString(values)
	builder.WriteString(")")
}

func buildFragmentsClause(query *sdlQuery, rp graphql.ResolveParams) {
	for len(query.fragmentsToAdd) > 0 {
		fragmentName := query.fragmentsToAdd[0]
		query.addedFragments[fragmentName] = true
		query.fragmentsToAdd = query.fragmentsToAdd[1:]
		fragment := rp.Info.Fragments[fragmentName].(*ast.FragmentDefinition)

		query.builder.WriteString(fmt.Sprintf("fragment %s on %s", fragmentName, fragment.TypeCondition.Name.Value))
		resolveParamsToQueryInner(query, fragment, rp)
	}
}

func buildVariableClause(query *sdlQuery) string {
	if len(query.variables) == 0 {
		return ""
	}

	builder := strings.Builder{}
	for varName, varType := range query.variables {
		builder.WriteString(fmt.Sprintf("$%s:%s,", varName, varType))
	}
	clause := builder.String()
	clause = clause[:len(clause)-1]
	return fmt.Sprintf("(%s)", clause)
}

func getVariableType(varName string, rp graphql.ResolveParams) string {
	variable := rp.Info.VariableValues[varName]

	switch variable.(type) {
	case int:
		return "Int"
	case string:
		return "String"
	default:
		return reflect.TypeOf(variable).Name()
	}
}

func getVariableValues(variables map[string]string, variableValues map[string]interface{}) map[string]interface{} {
	result := make(map[string]interface{})
	for varName := range variables {
		result[varName] = variableValues[varName]
	}
	return result
}
