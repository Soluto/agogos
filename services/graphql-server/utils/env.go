package utils

import "os"

func GetenvWithFallback(key, fallback string) string {
	res, ok := os.LookupEnv(key)
	if ok {
		return res
	}

	return fallback
}
