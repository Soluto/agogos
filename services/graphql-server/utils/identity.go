package utils

import "reflect"

func IdentityResolver(fieldName string, source interface{}) (res interface{}, err error) {
	defer Recovery(&err)

	switch source.(type) {

	case map[string]interface{}:
		m := source.(map[string]interface{})
		res = m[fieldName]

	default:
		value := reflect.ValueOf(source)
		f := reflect.Indirect(value).FieldByName(fieldName)
		res = f.Interface()
	}

	return
}
