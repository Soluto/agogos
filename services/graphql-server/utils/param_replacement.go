package utils

import (
	"fmt"
	"regexp"
	"strings"

	"github.com/graphql-go/graphql"
)

var reString = `{\w+\.\w+}`
var re = regexp.MustCompile(reString)
var reExactlyOneString = fmt.Sprintf("^%s$", reString)
var reExactlyOne = regexp.MustCompile(reExactlyOneString)

func resolveFromParams(rp graphql.ResolveParams, sourceName string, propName string) interface{} {
	switch sourceName {
	case "args":
		val, ok := rp.Args[propName]
		if !ok {
			return nil
		}
		return val
	case "source":
		res, err := IdentityResolver(propName, rp.Source)
		if err != nil {
			return nil
		}
		return res
	default:
		return nil
	}
}

// ReplaceWithParameters replaces template strings that look like "something {args.id} {source.someotherprop}", replacing those templates with the corresponding properties in the ResolveParams
func ReplaceWithParameters(rp graphql.ResolveParams, str string) string {
	return re.ReplaceAllStringFunc(str, func(s string) string {
		prop := getProp(rp, s)
		if prop == nil {
			return ""
		}

		return fmt.Sprintf("%v", prop)
	})
}

func getProp(rp graphql.ResolveParams, template string) interface{} {
	sepIndex := strings.IndexRune(template, '.')

	sourceName := template[1:sepIndex]
	propName := template[sepIndex+1 : len(template)-1]

	return resolveFromParams(rp, sourceName, propName)
}

// GetReplacementIfArray - If str is only a single template that resolves to an array, it returns that array, otherwise it returns nil
func GetReplacementIfArray(rp graphql.ResolveParams, str string) []interface{} {
	s := reExactlyOne.FindString(str)
	if s == "" {
		return nil
	}

	prop := getProp(rp, str)
	if prop == nil {
		return nil
	}

	res, ok := prop.([]interface{})
	if !ok {
		return nil
	}

	return res
}
