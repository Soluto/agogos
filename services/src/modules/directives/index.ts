import { concatAST } from 'graphql';
import { SchemaDirectiveVisitor } from 'graphql-tools';
import { sdl as stubSdl, StubDirective } from './stub';
import { sdl as localResolverSdl, LocalResolverDirective } from './local-resolver';
import { sdl as restSdl, RestDirective } from './rest';
import { sdl as gqlSdl, GqlDirective } from './gql';
import { sdl as exportSdl, ExportDirective } from './export';
import { sdl as selectSdl, SelectDirective } from './select';
import {
  policyBaseSdl,
  policySdl,
  policiesSdl,
  policyQuerySdl,
  PolicyDirective,
  PoliciesDirective,
  PolicyQueryDirective,
  policyScalarResolvers,
} from './policy';

export const directiveMap: { [visitorName: string]: typeof SchemaDirectiveVisitor } = {
  stub: StubDirective,
  rest: RestDirective,
  gql: GqlDirective,
  export: ExportDirective,
  select: SelectDirective,
  policy: PolicyDirective,
  policies: PoliciesDirective,
  policyQuery: PolicyQueryDirective,
  localResolver: LocalResolverDirective,
};

export const sdl = concatAST([
  stubSdl,
  restSdl,
  gqlSdl,
  exportSdl,
  selectSdl,
  policySdl,
  policiesSdl,
  policyQuerySdl,
  localResolverSdl,
]);

export { policyScalarResolvers, policyBaseSdl };
