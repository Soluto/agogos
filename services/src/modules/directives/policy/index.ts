import { PolicyDefinition } from '../../resource-repository';
import { sdl as policySdl, PolicyDirective } from './policy';
import { sdl as policiesSdl, PoliciesDirective, resolvers as policyScalarResolvers } from './policies';
import { sdl as policyQuerySdl, PolicyQueryDirective } from './policy-query';
import PolicyExecutor from './policy-executor';
import { Policy, LoadedPolicy } from './types';

export {
  PolicyExecutor,
  policySdl,
  policiesSdl,
  policyQuerySdl,
  PolicyDirective,
  PoliciesDirective,
  PolicyQueryDirective,
  policyScalarResolvers,
};

export type AuthorizationConfig = {
  policies: PolicyDefinition[];
  policyAttachments?: Record<string, LoadedPolicy>;
  policyExecutor: PolicyExecutor;
  basePolicy?: Policy;
};
