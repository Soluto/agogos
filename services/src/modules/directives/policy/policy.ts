import { GraphQLResolveInfo, GraphQLField, defaultFieldResolver, GraphQLObjectType, FieldNode } from 'graphql';
import { SchemaDirectiveVisitor } from 'graphql-tools';
import { gql } from 'apollo-server-core';
import { RequestContext } from '../../context';
import { GraphQLArguments, Policy } from './types';

export class PolicyDirective extends SchemaDirectiveVisitor {
  visitObject(object: GraphQLObjectType<unknown, RequestContext>) {
    const policy = this.args as Policy;
    const originalResolveObject = object.resolveObject;

    object.resolveObject = async (
      source: unknown,
      fields: Record<string, ReadonlyArray<FieldNode>>,
      context: RequestContext,
      info: GraphQLResolveInfo
    ) => {
      if (!context.ignorePolicies) {
        await context.authorizationConfig.policyExecutor.validatePolicy(policy, source, {}, context, info);
      }

      return originalResolveObject ? await originalResolveObject(source, fields, context, info) : source;
    };
  }

  visitFieldDefinition(field: GraphQLField<unknown, RequestContext>) {
    const policy = this.args as Policy;
    const originalResolve = field.resolve || defaultFieldResolver;

    field.resolve = async (
      source: unknown,
      args: GraphQLArguments,
      context: RequestContext,
      info: GraphQLResolveInfo
    ) => {
      if (!context.ignorePolicies) {
        await context.authorizationConfig.policyExecutor.validatePolicy(policy, source, args, context, info);
      }

      return originalResolve.call(field, source, args, context, info);
    };
  }
}

export const sdl = gql`
  directive @policy(namespace: String!, name: String!, args: JSONObject) on OBJECT | FIELD_DEFINITION
`;
