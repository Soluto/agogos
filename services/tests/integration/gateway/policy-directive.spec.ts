import { ApolloServerTestClient, createTestClient } from 'apollo-server-testing';
import { ApolloServerBase, gql } from 'apollo-server-core';
import { ApolloServer } from 'apollo-server-fastify';
import { IResolvers } from 'graphql-tools';
import { when } from 'jest-when';
import { concatAST, DocumentNode } from 'graphql';
import { PolicyExecutor } from '../../../src/modules/directives/policy';
import { sdl as lowerCaseSdl, LowerCaseDirective } from '../utils/lower-case-directive';
import { sdl as mockSdl, MockDirective } from '../utils/mock-directive';
import getBaseSchema, { BaseSchema } from '../../../src/modules/base-schema';
import GraphQLErrorSerializer from '../../utils/graphql-error-serializer';

const mockValidatePolicy = jest.fn();
// eslint-disable-next-line unicorn/no-useless-undefined
when(mockValidatePolicy).calledWith({ namespace: 'ns', name: 'alwaysAllow' }).mockResolvedValue(undefined);
when(mockValidatePolicy).calledWith({ namespace: 'ns', name: 'alwaysDeny' }).mockRejectedValue(new Error('Error'));
jest.mock('../../../src/modules/directives/policy/policy-executor', () => ({
  default: jest.fn().mockImplementation(() => ({ validatePolicy: mockValidatePolicy })),
}));

interface TestCase {
  typeDefs: DocumentNode;
  resolvers?: IResolvers;
}

const testCases: [string, TestCase][] = [
  [
    'Policy on field, resolver on object (ALLOW)',
    {
      typeDefs: gql`
        type Foo {
          bar: String! @policy(namespace: "ns", name: "alwaysAllow")
        }
        type Query {
          foo: Foo!
        }
      `,
      resolvers: {
        Query: {
          foo: () => ({ bar: 'BAR' }),
        },
      },
    },
  ],
  [
    'Policy on field, resolver on object (DENY)',
    {
      typeDefs: gql`
        type Foo {
          bar: String! @policy(namespace: "ns", name: "alwaysDeny")
        }
        type Query {
          foo: Foo!
        }
      `,
      resolvers: {
        Query: {
          foo: () => ({ bar: 'BAR' }),
        },
      },
    },
  ],
  [
    'Policy on field, stub on parent query (ALLOW)',
    {
      typeDefs: gql`
        type Foo {
          bar: String! @policy(namespace: "ns", name: "alwaysAllow")
        }
        type Query {
          foo: Foo! @localResolver(value: { bar: "BAR" })
        }
      `,
    },
  ],
  [
    'Policy on field, stub on parent query (DENY)',
    {
      typeDefs: gql`
        type Foo {
          bar: String! @policy(namespace: "ns", name: "alwaysDeny")
        }
        type Query {
          foo: Foo! @localResolver(value: { bar: "BAR" })
        }
      `,
    },
  ],
  [
    'Stub, policy on field (ALLOW)',
    {
      typeDefs: gql`
        type Foo {
          bar: String! @localResolver(value: "BAR") @policy(namespace: "ns", name: "alwaysAllow")
        }
        type Query {
          foo: Foo! @localResolver(value: {})
        }
      `,
    },
  ],
  [
    'Stub, policy on field (DENY)',
    {
      typeDefs: gql`
        type Foo {
          bar: String! @localResolver(value: "BAR") @policy(namespace: "ns", name: "alwaysDeny")
        }
        type Query {
          foo: Foo! @localResolver(value: {})
        }
      `,
    },
  ],
  [
    // This is incorrect order of directives so the policy doesn't work
    'Policy, stub on field (ALLOW)',
    {
      typeDefs: gql`
        type Foo {
          bar: String! @policy(namespace: "ns", name: "alwaysAllow") @localResolver(value: "BAR")
        }
        type Query {
          foo: Foo! @localResolver(value: {})
        }
      `,
    },
  ],
  [
    // This is incorrect order of directives so the policy doesn't work
    'Policy, stub on field (DENY)',
    {
      typeDefs: gql`
        type Foo {
          bar: String! @policy(namespace: "ns", name: "alwaysDeny") @localResolver(value: "BAR")
        }
        type Query {
          foo: Foo! @localResolver(value: {})
        }
      `,
    },
  ],
  [
    'Policy on object, stub on field (ALLOW)',
    {
      typeDefs: gql`
        type Foo @policy(namespace: "ns", name: "alwaysAllow") {
          bar: String! @localResolver(value: "BAR")
        }
        type Query {
          foo: Foo! @localResolver(value: {})
        }
      `,
    },
  ],
  [
    'Policy on object, stub on field (DENY)',
    {
      typeDefs: gql`
        type Foo @policy(namespace: "ns", name: "alwaysDeny") {
          bar: String! @localResolver(value: "BAR")
        }
        type Query {
          foo: Foo! @localResolver(value: {})
        }
      `,
    },
  ],
  [
    'Policy on object, stub on parent query (ALLOW)',
    {
      typeDefs: gql`
        type Foo @policy(namespace: "ns", name: "alwaysAllow") {
          bar: String!
        }
        type Query {
          foo: Foo! @localResolver(value: { bar: "BAR" })
        }
      `,
    },
  ],
  [
    'Policy, stub on parent query (DENY)',
    {
      typeDefs: gql`
        type Foo @policy(namespace: "ns", name: "alwaysDeny") {
          bar: String!
        }
        type Query {
          foo: Foo! @localResolver(value: { bar: "BAR" })
        }
      `,
    },
  ],
  [
    'Stub, lowerCase, policy on field (ALLOW)',
    {
      typeDefs: gql`
        type Foo {
          bar: String! @localResolver(value: "BAR") @lowerCase @policy(namespace: "ns", name: "alwaysAllow")
        }
        type Query {
          foo: Foo! @localResolver(value: {})
        }
      `,
    },
  ],
  [
    'Stub, lowerCase, policy on field (DENY)',
    {
      typeDefs: gql`
        type Foo {
          bar: String! @localResolver(value: "BAR") @lowerCase @policy(namespace: "ns", name: "alwaysDeny")
        }
        type Query {
          foo: Foo! @localResolver(value: {})
        }
      `,
    },
  ],
  [
    'Stub, policy, lowerCase on field (ALLOW)',
    {
      // This is incorrect order of directives but it still does work by chance.
      typeDefs: gql`
        type Foo {
          bar: String! @localResolver(value: "BAR") @policy(namespace: "ns", name: "alwaysAllow") @lowerCase
        }
        type Query {
          foo: Foo! @localResolver(value: {})
        }
      `,
    },
  ],
  [
    // This is incorrect order of directives but it still does work by chance.
    'Stub, policy, lowerCase on field (DENY)',
    {
      typeDefs: gql`
        type Foo {
          bar: String! @localResolver(value: "BAR") @policy(namespace: "ns", name: "alwaysDeny") @lowerCase
        }
        type Query {
          foo: Foo! @localResolver(value: {})
        }
      `,
    },
  ],
  [
    'LowerCase on object, policy on field, stub on parent query (ALLOW)',
    {
      typeDefs: gql`
        type Foo @lowerCase {
          bar: String! @policy(namespace: "ns", name: "alwaysAllow")
        }
        type Query {
          foo: Foo! @localResolver(value: { bar: "BAR" })
        }
      `,
    },
  ],
  [
    'LowerCase on object, policy on field, stub on parent query (DENY)',
    {
      typeDefs: gql`
        type Foo @lowerCase {
          bar: String! @policy(namespace: "ns", name: "alwaysDeny")
        }
        type Query {
          foo: Foo! @localResolver(value: { bar: "BAR" })
        }
      `,
    },
  ],
  [
    'LowerCase, policy on object, stub on parent query (ALLOW)',
    {
      typeDefs: gql`
        type Foo @lowerCase @policy(namespace: "ns", name: "alwaysAllow") {
          bar: String!
        }
        type Query {
          foo: Foo! @localResolver(value: { bar: "BAR" })
        }
      `,
    },
  ],
  [
    'LowerCase, policy on object, stub on parent query (DENY)',
    {
      typeDefs: gql`
        type Foo @lowerCase @policy(namespace: "ns", name: "alwaysDeny") {
          bar: String!
        }
        type Query {
          foo: Foo! @localResolver(value: { bar: "BAR" })
        }
      `,
    },
  ],
  [
    // This is incorrect order of directives so the policy doesn't work
    'Policy, lowerCase on object, stub on parent query (ALLOW)',
    {
      typeDefs: gql`
        type Foo @policy(namespace: "ns", name: "alwaysAllow") @lowerCase {
          bar: String!
        }
        type Query {
          foo: Foo! @localResolver(value: { bar: "BAR" })
        }
      `,
    },
  ],
  [
    // This is incorrect order of directives so the policy doesn't work
    'Policy, lowerCase on object, stub on parent query (DENY)',
    {
      typeDefs: gql`
        type Foo @policy(namespace: "ns", name: "alwaysDeny") @lowerCase {
          bar: String!
        }
        type Query {
          foo: Foo! @localResolver(value: { bar: "BAR" })
        }
      `,
    },
  ],
  [
    'Mock, policy on object, stub on parent query (ALLOW)',
    {
      typeDefs: gql`
        type Foo @mock @policy(namespace: "ns", name: "alwaysAllow") {
          bar: String!
        }
        type Query {
          foo: Foo! @localResolver(value: { id: "1" })
        }
      `,
    },
  ],
  [
    'Mock, policy on object, stub on parent query (DENY)',
    {
      typeDefs: gql`
        type Foo @mock @policy(namespace: "ns", name: "alwaysDeny") {
          bar: String!
        }
        type Query {
          foo: Foo! @localResolver(value: { id: "1" })
        }
      `,
    },
  ],
];

describe('Policy Directive Tests', () => {
  let client: ApolloServerTestClient;
  let server: ApolloServerBase;

  const query = gql`
    query {
      foo {
        bar
      }
    }
  `;

  let baseSchema: BaseSchema;

  beforeAll(async () => {
    baseSchema = await getBaseSchema();
    expect.addSnapshotSerializer(GraphQLErrorSerializer);
  });

  test.each(testCases)('%s', async (_, { typeDefs, resolvers }) => {
    server = new ApolloServer({
      typeDefs: concatAST([baseSchema.typeDefs, typeDefs, lowerCaseSdl, mockSdl]),
      resolvers: [resolvers ?? {}, baseSchema.resolvers],
      schemaDirectives: {
        ...baseSchema.directives,
        lowerCase: LowerCaseDirective,
        mock: MockDirective,
      },
      context: {
        authorizationConfig: {
          policyExecutor: new PolicyExecutor(),
        },
      },
    });
    client = createTestClient(server);

    const response = await client.query({ query });
    expect(response).toMatchSnapshot();
  });
});
