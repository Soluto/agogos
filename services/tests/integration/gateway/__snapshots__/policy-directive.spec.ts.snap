// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Policy Directive Tests LowerCase on object, policy on field, stub on parent query (ALLOW) 1`] = `
Object {
  "data": Object {
    "foo": Object {
      "bar": "bar",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Policy Directive Tests LowerCase on object, policy on field, stub on parent query (DENY) 1`] = `
{
  "errors": [
    {
      "name": "GraphQLError",
      "message": "Error",
      "exception": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    }
  ],
  "data": {}
}
`;

exports[`Policy Directive Tests LowerCase, policy on object, stub on parent query (ALLOW) 1`] = `
Object {
  "data": Object {
    "foo": Object {
      "bar": "bar",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Policy Directive Tests LowerCase, policy on object, stub on parent query (DENY) 1`] = `
{
  "errors": [
    {
      "name": "GraphQLError",
      "message": "Error",
      "exception": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    }
  ],
  "data": {}
}
`;

exports[`Policy Directive Tests Mock, policy on object, stub on parent query (ALLOW) 1`] = `
Object {
  "data": Object {
    "foo": Object {
      "bar": "BAR: 1",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Policy Directive Tests Mock, policy on object, stub on parent query (DENY) 1`] = `
{
  "errors": [
    {
      "name": "GraphQLError",
      "message": "Error",
      "exception": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    }
  ],
  "data": {}
}
`;

exports[`Policy Directive Tests Policy on field, resolver on object (ALLOW) 1`] = `
Object {
  "data": Object {
    "foo": Object {
      "bar": "BAR",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Policy Directive Tests Policy on field, resolver on object (DENY) 1`] = `
{
  "errors": [
    {
      "name": "GraphQLError",
      "message": "Error",
      "exception": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    }
  ],
  "data": {}
}
`;

exports[`Policy Directive Tests Policy on field, stub on parent query (ALLOW) 1`] = `
Object {
  "data": Object {
    "foo": Object {
      "bar": "BAR",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Policy Directive Tests Policy on field, stub on parent query (DENY) 1`] = `
{
  "errors": [
    {
      "name": "GraphQLError",
      "message": "Error",
      "exception": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    }
  ],
  "data": {}
}
`;

exports[`Policy Directive Tests Policy on object, stub on field (ALLOW) 1`] = `
Object {
  "data": Object {
    "foo": Object {
      "bar": "BAR",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Policy Directive Tests Policy on object, stub on field (DENY) 1`] = `
{
  "errors": [
    {
      "name": "GraphQLError",
      "message": "Error",
      "exception": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    }
  ],
  "data": {}
}
`;

exports[`Policy Directive Tests Policy on object, stub on parent query (ALLOW) 1`] = `
Object {
  "data": Object {
    "foo": Object {
      "bar": "BAR",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Policy Directive Tests Policy, lowerCase on object, stub on parent query (ALLOW) 1`] = `
Object {
  "data": Object {
    "foo": Object {
      "bar": "bar",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Policy Directive Tests Policy, lowerCase on object, stub on parent query (DENY) 1`] = `
{
  "errors": [
    {
      "name": "GraphQLError",
      "message": "Error",
      "exception": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    }
  ],
  "data": {}
}
`;

exports[`Policy Directive Tests Policy, stub on field (ALLOW) 1`] = `
Object {
  "data": Object {
    "foo": Object {
      "bar": "BAR",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Policy Directive Tests Policy, stub on field (DENY) 1`] = `
Object {
  "data": Object {
    "foo": Object {
      "bar": "BAR",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Policy Directive Tests Policy, stub on parent query (DENY) 1`] = `
{
  "errors": [
    {
      "name": "GraphQLError",
      "message": "Error",
      "exception": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    }
  ],
  "data": {}
}
`;

exports[`Policy Directive Tests Stub, lowerCase, policy on field (ALLOW) 1`] = `
Object {
  "data": Object {
    "foo": Object {
      "bar": "bar",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Policy Directive Tests Stub, lowerCase, policy on field (DENY) 1`] = `
{
  "errors": [
    {
      "name": "GraphQLError",
      "message": "Error",
      "exception": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    }
  ],
  "data": {}
}
`;

exports[`Policy Directive Tests Stub, policy on field (ALLOW) 1`] = `
Object {
  "data": Object {
    "foo": Object {
      "bar": "BAR",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Policy Directive Tests Stub, policy on field (DENY) 1`] = `
{
  "errors": [
    {
      "name": "GraphQLError",
      "message": "Error",
      "exception": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    }
  ],
  "data": {}
}
`;

exports[`Policy Directive Tests Stub, policy, lowerCase on field (ALLOW) 1`] = `
Object {
  "data": Object {
    "foo": Object {
      "bar": "bar",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Policy Directive Tests Stub, policy, lowerCase on field (DENY) 1`] = `
{
  "errors": [
    {
      "name": "GraphQLError",
      "message": "Error",
      "exception": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    }
  ],
  "data": {}
}
`;
